internals-js/src/specs/sourceSpec.ts

```diff
@@ -140,8 +144,271 @@ export class SourceSpecDefinition extends FeatureDefinition {
+          if (rest[knownProtocol]) {
+            if (protocol) {
+              errors.push(new GraphQLError(
+                `${sourceAPIInSchema.name} must specify only one of ${KNOWN_SOURCE_PROTOCOLS.join(', ')}`,
+              ));
+            }
+            protocol = knownProtocol;
+          }
+        });
+
+        if (protocol) {
+          apiNameToProtocol.set(name, protocol);
+
+          const protocolValue = rest[protocol];
+          if (protocolValue && protocol === "http") {
+            const { baseURL, headers } = protocolValue as HTTPSourceAPI;
+
+            try {
+              new URL(baseURL);
+            } catch (e) {
+              errors.push(new GraphQLError(`${sourceAPIInSchema.name} http.baseURL ${
+                JSON.stringify(baseURL)
+              } must be valid URL`));
+            }
+
+            validateHTTPHeaders(headers, errors, sourceAPIInSchema.name);
+          }
+        } else {
+          errors.push(new GraphQLError(
+            `${sourceAPIInSchema.name} must specify one of ${KNOWN_SOURCE_PROTOCOLS.join(', ')}`,
+          ));
+        }
+      });
+    }
+
+    if (sourceTypeInSchema) {
+      sourceTypeInSchema.applications().forEach(application => {
+        const { api, ...rest } = application.arguments();
+        if (!api || !apiNameToProtocol.has(api)) {
+          errors.push(new GraphQLError(`${sourceTypeInSchema.name} specifies unknown api ${api}`));
+        } else {
+          const expectedProtocol = apiNameToProtocol.get(api);
+          const protocolValue = expectedProtocol && rest[expectedProtocol];
+          if (expectedProtocol && !protocolValue) {
+            errors.push(new GraphQLError(
+              `${sourceTypeInSchema.name} must specify same ${
+                expectedProtocol
+              } argument as corresponding ${
+                sourceAPIInSchema!.name
+              } for api ${api}`,
+            ));
+          }
+
+          if (protocolValue && expectedProtocol === "http") {
+            const { GET, POST, headers, body } = protocolValue as HTTPSourceType;
+
+            if ([GET, POST].filter(Boolean).length !== 1) {
+              errors.push(new GraphQLError(
+                `${sourceTypeInSchema.name} must specify exactly one of http.GET or http.POST`,
+              ));
+            } else {
+              const urlPathTemplate = (GET || POST)!;
+              try {
+                // TODO Validate URL path template uses only available fields of
+                // the type.
+                parseURLPathTemplate(urlPathTemplate);
+              } catch (e) {
+                errors.push(new GraphQLError(
+                  `${sourceTypeInSchema.name} http.GET or http.POST must be valid URL path template`,
+                ));
+              }
+            }
+
+            validateHTTPHeaders(headers, errors, sourceTypeInSchema.name);
+
+            if (body) {
+              try {
+                parseJSONSelection(body);
```

Comment:
Sorry to be dumb, but how does `body` within the `HTTPSourceField` relate to `selection` on `sourceType`? We will want to validate that field as well.

Reply:
Both `http.body` and the `selection` arguments are `JSONSelection` strings. The `http.body` selector for `@sourceField` produces an HTTP request body from the field's arguments, whereas the `selection` argument processes JSON received from the REST API so it matches the desired GraphQL fields. In other words, the inputs are different, but the selection syntax is shared.